### DISCUSSION TITLE
The MQTTClient_connectionLost callback is always called without warning

### DISCUSSION THREAD

**Discussion by hiwooiot** [eclipse-paho/paho.mqtt.c#1407] [General]
Data is being sent every second and can be subscribed to, but the mqtt connection will suddenly disconnect periodically.  I don't know what caused it.  My understanding is that the data can always be sent successfully and even if the PINGREQ fails, the connection should not be disconnected.

20231013 143302.527 346492 BoxPlugIn5200023062019300033 -> PUBLISH qos: 0 retained: 0 rc: -22 payload len(54209): {"deviceLogData":[{"
20231013 143302.527 Storing unsent QoS 0 message
20231013 143303.044 Return code 0 from read select
20231013 143303.044 Return code 0 from write select
20231013 143303.044 Return code 0 from read select
20231013 143303.044 Return code 0 from write select
20231013 143303.044 Return code 1 from read select
20231013 143303.044 ContinueWrite: partial write now complete for socket 346492
20231013 143303.044 Return code 0 from write select
20231013 143303.044 Partial write: 0 bytes of 54291 actually written on socket 346492
20231013 143303.044 346492 BoxPlugIn5200023062019300033 -> PUBLISH qos: 0 retained: 0 rc: -22 payload len(54248): {"deviceLogData":[{"
20231013 143303.044 Storing unsent QoS 0 message
20231013 143303.044 Return code 1 from read select
20231013 143303.044 ContinueWrite: partial write now complete for socket 346492
20231013 143303.044 Return code 0 from write select
20231013 143303.044 Partial write: 0 bytes of 54291 actually written on socket 346492
20231013 143303.044 346492 BoxPlugIn5200023062019300033 -> PUBLISH qos: 0 retained: 0 rc: -22 payload len(54248): {"deviceLogData":[{"
20231013 143303.044 Storing unsent QoS 0 message
20231013 143303.044 Return code 0 from read select
20231013 143303.044 Return code 0 from write select
20231013 143303.044 Return code 1 from read select
20231013 143303.044 ContinueWrite: partial write now complete for socket 346492
20231013 143303.538 Return code 0 from write select
20231013 143303.538 Partial write: 0 bytes of 54291 actually written on socket 346492
20231013 143303.538 346492 BoxPlugIn5200023062019300033 -> PUBLISH qos: 0 retained: 0 rc: -22 payload len(54248): {"deviceLogData":[{"
20231013 143303.538 Storing unsent QoS 0 message
20231013 143303.538 Return code 0 from read select
20231013 143303.538 Return code 0 from write select
20231013 143303.538 Return code 1 from read select
20231013 143303.538 ContinueWrite: partial write now complete for socket 346492
20231013 143303.538 Return code 0 from write select
20231013 143303.538 Partial write: 0 bytes of 54291 actually written on socket 346492
20231013 143303.538 346492 BoxPlugIn5200023062019300033 -> PUBLISH qos: 0 retained: 0 rc: -22 payload len(54248): {"deviceLogData":[{"
20231013 143303.538 Storing unsent QoS 0 message
20231013 143303.538 Return code 0 from read select
20231013 143303.538 Return code 0 from write select
20231013 143303.538 Return code 1 from read select
20231013 143303.538 ContinueWrite: partial write now complete for socket 346492
20231013 143303.538 Return code 0 from write select
20231013 143303.538 Partial write: 0 bytes of 2 actually written on socket 346492
20231013 143303.538 346492 BoxPlugIn5200023062019300033 -> PINGREQ (-22)
20231013 143303.538 Error sending PINGREQ for client BoxPlugIn5200023062019300033 on socket 346492, disconnecting
20231013 143303.915 Trying to write to socket 346492 for which there is already pending output
20231013 143303.915 346492 BoxPlugIn5200023062019300033 -> DISCONNECT (-1)
20231013 143303.915 Cleaning in abortWrite for socket 346492
20231013 143303.915 Removed socket 346492
20231013 143303.915 Reset max fdp1 to 1
20231013 143303.915 Conn_count is 0

**Comment by hiwooiot** [comment]
I am using the latest paho mqtt 1.2.12 version of the library, which runs in a windows environment

**Comment by icraggs** [comment]
That looks like an old error message to me - what version did you actually mean (1.2.12 doesn't exist).

You don't show the beginning of the trace which makes it harder to see what is going on. But it looks like you might be trying to send more data than your connection can handle, or at least very close to the limit. I would try sending the message less frequently or reduce the size of the payload and see if that makes any difference, to start with.

**Reply by hiwooiot** [reply]
Can MQTTClient_publishMessage be called in the MQTTClient_messageArrived callback function？

**Reply by hiwooiot** [reply]
Version 1.3.12, after the packet size of each published message was changed from about 50kb to about 10kb, there were link interruptions but not so frequently, I checked the log and found that the link was immediately disconnected when the socket was interrupted when the PINGREQ message was sent.  This causes the data to be sent normally but always suddenly disconnected.
I am running on windows7 and the Internet speed is not very good (180kb/s), which may be the cause of the break, but I do not understand why the link always breaks when the data is sent normally, without any warning

**Comment by icraggs** [comment]
> Can MQTTClient_publishMessage be called in the MQTTClient_messageArrived callback function？

No. In the MQTTClient API, the only call that can be made in a callback is MQTTClient_connect in the connection lost callback. If you want to do this, you have to use the MQTTAsync API.

**Comment by icraggs** [comment]
If you change the QoS of the message to 1, then the library won't need to send any pings, because it will be receiving ACK messages back from the broker. That won't help if you are still trying to send more data than the TCP connection can handle however. 

You can also turn off keepalive (sending PINGs) by setting the keepalive interval to 0.

**Reply by hiwooiot** [reply]
Thank you very much, I ran one night after QOS was changed to 1 without any broken links
